### PR TITLE
Created a REST API that other services can make requests to in order to create notifications

### LINKED ISSUE #53: Major: Need to add notification events on liking posts.
Currently no functionality exists for notifications being sent on someone liking a post.

### LINKED ISSUE #54: Major: Need to add notification events for commenting on posts.
Currently no functionality exists for sending notification events when someone comments on a post.

### LINKED ISSUE #58: Add REST API for other back-end services to trigger notifications
this is in attempt to resolve the following issues:

#56 
#54 
#53

### PR DESCRIPTION
**Pull request by ibnetahir1**
- Created NotificationController in WebApi project that has 3 action methods for creating a comment/like/follow notification
- Other backend services should access this API in order to create their respective notifications and send those to the clients. This removes the need for the frontend to send notification requests.

### CODE REVIEW

**Review by jpw547-zz — COMMENTED** [review]
The tests look alright to me. Quality gate seems bugged...

File: FakebookNotifications/FakebookNotifications.Testing/IntegrationTests/NotificationControllerTest.cs [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+            mockGroups.Setup(group => group.AddToGroupAsync(It.IsIn<string>(clientIds), It.IsIn<string>(groupIds), new CancellationToken())).Returns(Task.FromResult(true));
+            mockGroups.Setup(group => group.RemoveFromGroupAsync(It.IsIn<string>(clientIds), It.IsIn<string>(groupIds), new CancellationToken())).Returns(Task.FromResult(true));
+            mockContext.Setup(context => context.ConnectionId).Returns("01");
+            
+
+            _logger = new NullLogger<NotificationsContext>();
+            Mock<IOptions<NotificationsDatabaseSettings>> mockSettings = new();
+            settings = new NotificationsDatabaseSettings
+            {
+                ConnectionString = "mongodb+srv://ryan:1234@fakebook.r8oce.mongodb.net/Notifications?retryWrites=true&w=majority",
```
Comment (escalonn) [inline FakebookNotifications/FakebookNotifications.Testing/IntegrationTests/NotificationControllerTest.cs]: don't send your database connection string to github, even your test database. remove this string & change the password on it.
if we want a test database, better to run it with docker in the pipeline. we'd want a "trait" on all tests that required that, so we could skip running those tests when we don't want to run the test db.
but you're confused on what you're doing here. if you're mocking the mongodb driver, writing a unit test, then there is zero purpose to having a real database. if you're writing an integration test with a real db, then there's no purpose to mocking mongodb driver. figure out what kind of tests you want, and if you want both kinds, put them in separate namespaces at least.

File: FakebookNotifications/FakebookNotifications.Testing/IntegrationTests/NotificationControllerTest.cs [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+                //ConnectionString = "mongodb://localhost:27017",
+                DatabaseName = "Notifications",
+                UserCollection = "User",
+                NotificationsCollection = "Notifications"
+            };
+            mockSettings.Setup(s => s.Value).Returns(settings);
+            NotificationsContext mockDbContext = new(mockSettings.Object, _logger);
```
Comment (escalonn) [inline FakebookNotifications/FakebookNotifications.Testing/IntegrationTests/NotificationControllerTest.cs]: it's not a mock if you're calling the real constructor. figure out what you're trying to do

File: FakebookNotifications/FakebookNotifications.Testing/IntegrationTests/NotificationControllerTest.cs [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+        private readonly Mock<IHubCallerClients> mockClients = new();
+        private readonly Mock<IGroupManager> mockGroups = new();
+        private readonly Mock<IClientProxy> mockClientProxy = new();
+        private readonly Mock<HubCallerContext> mockContext = new();
+        private readonly NotificationHub mockHub;
+
+        private readonly UserRepo _userRepo;
+        private readonly NotificationsRepo _noteRepo;
+        private readonly NotificationsDatabaseSettings settings;
+        private readonly NullLogger<NotificationsContext> _logger;
```
Comment (escalonn) [inline FakebookNotifications/FakebookNotifications.Testing/IntegrationTests/NotificationControllerTest.cs]: use a consistent naming convention for private fields - in this project, that's an initial underscore

File: FakebookNotifications/FakebookNotifications.Testing/IntegrationTests/NotificationControllerTest.cs [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+                DatabaseName = "Notifications",
+                UserCollection = "User",
+                NotificationsCollection = "Notifications"
+            };
+            mockSettings.Setup(s => s.Value).Returns(settings);
+            NotificationsContext mockDbContext = new(mockSettings.Object, _logger);
+
+            // mocking Mongo db
+            NotificationsRepo noteRepo = new(mockDbContext);
+            UserRepo userRepo = new(mockDbContext, noteRepo);
```
Comment (escalonn) [inline FakebookNotifications/FakebookNotifications.Testing/IntegrationTests/NotificationControllerTest.cs]: the integration test includes these objects too? yeah, i guess you could write an integration test that spans all those classes. but it's certainly not mocking.

File: FakebookNotifications/FakebookNotifications.WebApi/Controller/NotificationController.cs [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    [ApiController]
+    [Route("api/notification")]
+
+    public class NotificationController : ControllerBase
+    {
+        private readonly IHubContext<NotificationHub> hub;
+        private readonly INotificationsRepo notificationsRepo;
+        private readonly IUserRepo userRepo;
+
+        public NotificationController(IHubContext<NotificationHub> _hub, INotificationsRepo _notificationsRepo, IUserRepo _userRepo)
```
Comment (escalonn) [inline FakebookNotifications/FakebookNotifications.WebApi/Controller/NotificationController.cs]: inverted naming convention. parameters are camelCase, private fields are _camelCase

File: FakebookNotifications/FakebookNotifications.WebApi/Controller/NotificationController.cs [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+                foreach (var connection in user.Connections)
+                {
+                    await hub.Groups.AddToGroupAsync(connection, user.Email);
+                }
+
+                await hub.Clients.Group(user.Email).SendAsync(methodName, notification);
+                return Ok();
+            }
+            catch (Exception e)
```
Comment (escalonn) [inline FakebookNotifications/FakebookNotifications.WebApi/Controller/NotificationController.cs]: not every exception is the fault of the client. unexpected exceptions are generally the server's fault.

this exception handling is actively bad in worse ways though: you're preventing the default logging of uncaught exceptions, so it's harder for the devs to detect an issue, while sending the exception to the client, so it's easy for hackers to detect an issue.

File: FakebookNotifications/FakebookNotifications.WebApi/Controller/NotificationController.cs [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+            {
+                await notificationsRepo.CreateNotificationAsync(notification);
+                User user = await userRepo.GetUserAsync(loggedInUser);
+
+                foreach (var connection in user.Connections)
+                {
+                    await hub.Groups.AddToGroupAsync(connection, user.Email);
+                }
+
+                await hub.Clients.Group(user.Email).SendAsync(methodName, notification);
```
Comment (escalonn) [inline FakebookNotifications/FakebookNotifications.WebApi/Controller/NotificationController.cs]: i don't know signalr but are you sure this is right? you have to rebuild a group out of a user's connections every time just to send him something? and you have to do it slowly, one by one? sometimes it's unavoidable but you usually want to avoid `await` inside a loop at all costs

File: FakebookNotifications/FakebookNotifications.WebApi/Hubs/NotificationHub.cs [hunk 8]
```diff
@@ -15,7 +15,7 @@ namespace FakebookNotifications.WebApi.Hubs
     [Authorize]
     public class NotificationHub : Hub
     {
-
+        public static IHubContext<NotificationHub> Current { get; set; }
```
Comment (escalonn) [inline FakebookNotifications/FakebookNotifications.WebApi/Hubs/NotificationHub.cs]: what's this for?

File: FakebookNotifications/FakebookNotifications.WebApi/Startup.cs [hunk 9]
```diff
@@ -34,6 +34,7 @@ public void ConfigureServices(IServiceCollection services)
             services.AddScoped<INotificationsContext, NotificationsContext>();
             services.AddScoped<IUserRepo, UserRepo>();
             services.AddScoped<INotificationsRepo, NotificationsRepo>();
+
```
Comment (escalonn) [inline FakebookNotifications/FakebookNotifications.WebApi/Startup.cs:37]: unnecessary diff

**Review by escalonn — CHANGES_REQUESTED** [review]
fyi, async void is ordinarily bad (look up why). for xunit test methods though, it doesn't matter, so you could have marked as false positive. on the other hand, if we keep consistently to async Task, there are no downsides and future code won't have to manually review those issues.

File: FakebookNotifications/FakebookNotifications.WebApi/Controller/NotificationController.cs (hunk 7 above)
Comment (ibnetahir1) [reply FakebookNotifications/FakebookNotifications.WebApi/Controller/NotificationController.cs]: The way it is currently set up, we have to do it this way. This sends the notification to all different places the same user is connected from (laptop, mobile, etc.).

File: FakebookNotifications/FakebookNotifications.WebApi/Hubs/NotificationHub.cs (hunk 8 above)
Comment (ibnetahir1) [reply FakebookNotifications/FakebookNotifications.WebApi/Hubs/NotificationHub.cs]: That was a mistake

File: FakebookNotifications/FakebookNotifications.Testing/IntegrationTests/NotificationControllerTest.cs [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+            Connections = new List<string>{
+                "00", "02", "03"
+            },
+            Email = "test@test.com"
+        };
+
+        private readonly Mock<IGroupManager> _mockGroups = new();
+        private readonly Mock<IClientProxy> _mockClientProxy = new();
+        private readonly Mock<IUserRepo> _userRepo = new();
+        private readonly Mock<INotificationsRepo> _noteRepo = new();
```
Comment (escalonn) [inline FakebookNotifications/FakebookNotifications.Testing/IntegrationTests/NotificationControllerTest.cs]: name these mocks consistently.

File: FakebookNotifications/FakebookNotifications.Testing/IntegrationTests/NotificationControllerTest.cs [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+using Microsoft.Extensions.Logging.Abstractions;
+using Microsoft.Extensions.Options;
+using Moq;
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace FakebookNotifications.Testing.IntegrationTests
```
Comment (escalonn) [inline FakebookNotifications/FakebookNotifications.Testing/IntegrationTests/NotificationControllerTest.cs]: these are unit tests, not integration tests, so move them to a different namespace & folder

File: FakebookNotifications/FakebookNotifications.Testing/IntegrationTests/NotificationControllerTest.cs [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+            var postId = 0;
+
+            // act
+            var result = await controller.CommentNotificationAsync(dummy.LoggedInUserId, dummy.TriggerUserId, postId);
+
+            // assert
+            Assert.NotNull(result);
+            _noteRepo.Verify(c => c.CreateNotificationAsync(It.IsAny<Domain.Models.Notification>()), Times.Once);
+            _userRepo.Verify(c => c.GetUserAsync(It.IsAny<string>()), Times.Once);
+            _mockGroups.Verify(c => c.AddToGroupAsync(It.IsAny<string>(), It.IsAny<string>(), new CancellationToken()), Times.Exactly(3));
```
Comment (escalonn) [inline FakebookNotifications/FakebookNotifications.Testing/IntegrationTests/NotificationControllerTest.cs]: why are you passing in a cancellationtoken in these tests

File: FakebookNotifications/FakebookNotifications.Testing/IntegrationTests/NotificationControllerTest.cs [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+            Assert.IsAssignableFrom<IActionResult>(result);
+            Assert.IsType<OkResult>(result);
```
Comment (escalonn) [inline FakebookNotifications/FakebookNotifications.Testing/IntegrationTests/NotificationControllerTest.cs]: these three assertions on result are redundant in a few ways. the compiler can already guarantee that it's an iactionresult so that call does nothing. and the final IsType call makes the IsAssignableFrom and NotNull calls redundant too.
also, to follow liskov substitution principle, don't require the exact type. so assert `Assert.IsAssignableFrom<OkResult>(result);`, and it replaces the other assertions too. likewise for the similar tests.

File: FakebookNotifications/FakebookNotifications.WebApi/Controller/NotificationController.cs (hunk 5 above)
Comment (escalonn) [reply FakebookNotifications/FakebookNotifications.WebApi/Controller/NotificationController.cs]: previous review comment unaddressed.

File: FakebookNotifications/FakebookNotifications.WebApi/Controller/NotificationController.cs [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+        /// API call from other services to create a comment notification
+        /// </summary>
+        /// <param name="loggedInUser">email of the user who owns the post</param>
+        /// <param name="triggerUser">email of user who commented</param>
+        /// <param name="postId">Post ID of the post that was commented on</param>
+        /// <returns>Ok if accepted, BadRequest if fail</returns>
+        [HttpPost("comment")]
+        public async Task<IActionResult> CommentNotificationAsync(string loggedInUser, string triggerUser, int postId)
+        {
+            return await CreateNotificationAsync(loggedInUser, triggerUser, postId, "comment");
```
Comment (escalonn) [inline FakebookNotifications/FakebookNotifications.WebApi/Controller/NotificationController.cs]: at least use constants, not magic strings

File: FakebookNotifications/FakebookNotifications.WebApi/Controller/NotificationController.cs [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+                methodName = "CommentNotification";
+            }
+            else if (notificationType == "like")
+            {
+                methodName = "LikeNotification";
+            }
+            else if (notificationType == "follow")
+            {
+                methodName = "FollowNotification";
+            }
```
Comment (escalonn) [inline FakebookNotifications/FakebookNotifications.WebApi/Controller/NotificationController.cs]: no exception if you pass an invalid argument? not good for maintainability.

File: FakebookNotifications/FakebookNotifications.WebApi/Startup.cs (hunk 9 above)
Comment (escalonn) [reply FakebookNotifications/FakebookNotifications.WebApi/Startup.cs:37]: previous review comment unaddressed.

File: FakebookNotifications/FakebookNotifications.Testing/IntegrationTests/NotificationRepoTests.cs [hunk 16]
```diff
@@ -132,12 +132,29 @@ public async Task DeleteNotification_RepoTest()
...
+                HasBeenRead = true,
+                Date = DateTime.Now
+            };
+
+            //Act
+            var result = await repo.CreateNotificationAsync(notification);
+
+            //Assert
+            Assert.True(result);
+
```
Comment (escalonn) [inline FakebookNotifications/FakebookNotifications.Testing/IntegrationTests/NotificationRepoTests.cs:148]: why are we adding this? at a glance i guess this is an improvement on the existing test... not sure how it was passing before. at least change the test name and the arrange/act/assert comments to reflect what's actually going on.

these tests should be refactored eventually

**Review by escalonn — CHANGES_REQUESTED** [review]

**Comment by escalonn** [comment]
also click 'resolve conversation' on review comments when you push commits that you think resolve them

and just in case this comment gets lost among the older items on this PR - https://github.com/2011-fakebook-project3/notifications/pull/63#discussion_r611923924

File: FakebookNotifications/FakebookNotifications.WebApi/Controller/NotificationController.cs (hunk 7 above)
Comment (escalonn) [reply FakebookNotifications/FakebookNotifications.WebApi/Controller/NotificationController.cs]: @ibnetahir1 at least add some comments giving some context that would help future development work to possibly improve this aspect. maybe reference links if you based your code/design on something you read online.

File: FakebookNotifications/FakebookNotifications.WebApi/Startup.cs (hunk 9 above)
Comment (ibnetahir1) [reply FakebookNotifications/FakebookNotifications.WebApi/Startup.cs:37]: I believe we were trying to add some things, but then deleted the changes so that one extra line stayed. I guess the right call would have been to restore the file.

File: FakebookNotifications/FakebookNotifications.Testing/IntegrationTests/NotificationControllerTest.cs (hunk 12 above)
Comment (ibnetahir1) [reply FakebookNotifications/FakebookNotifications.Testing/IntegrationTests/NotificationControllerTest.cs]: It would not allow me to use the verify function unless I passed something for all parameters.

File: FakebookNotifications/FakebookNotifications.Testing/IntegrationTests/NotificationRepoTests.cs (hunk 16 above)
Comment (ibnetahir1) [reply FakebookNotifications/FakebookNotifications.Testing/IntegrationTests/NotificationRepoTests.cs:148]: I removed the unnecessary check.

File: FakebookNotifications/FakebookNotifications.WebApi/Controller/NotificationController.cs [hunk 17]
```diff
@@ -99,17 +102,22 @@ private async Task<IActionResult> CreateNotificationAsync(string loggedInUser, s
             {
                 methodName = "FollowNotification";
             }
+            else
+            {
+                throw new ArgumentException("notificationType is not valid");
```
Comment (escalonn) [inline FakebookNotifications/FakebookNotifications.WebApi/Controller/NotificationController.cs]: use the paramName argument to this constructor as well

File: FakebookNotifications/FakebookNotifications.WebApi/Controller/NotificationController.cs [hunk 18]
```diff
@@ -34,7 +34,8 @@ public NotificationController(IHubContext<NotificationHub> _hub, INotificationsR
         [HttpPost("comment")]
         public async Task<IActionResult> CommentNotificationAsync(string loggedInUser, string triggerUser, int postId)
         {
-            return await CreateNotificationAsync(loggedInUser, triggerUser, postId, "comment");
+            string notificationType = "comment";
```
Comment (escalonn) [inline FakebookNotifications/FakebookNotifications.WebApi/Controller/NotificationController.cs]: I meant declare the string "comment" once, and reference that one location both from methods like this and from the CreateNotificationAsync method. that will let the compiler catch typos in the notification type. i guess an enum would be better - you can make it a private nested enum

**Review by escalonn — CHANGES_REQUESTED** [review]

File: FakebookNotifications/FakebookNotifications.WebApi/Controller/NotificationController.cs (hunk 17 above)
Comment (escalonn) [reply FakebookNotifications/FakebookNotifications.WebApi/Controller/NotificationController.cs]: and use the new "switch expression" syntax, it's a good fit here

**Review by escalonn — APPROVED** [review]

File: FakebookNotifications/FakebookNotifications.WebApi/Controller/NotificationController.cs (hunk 17 above)
Comment (escalonn) [reply FakebookNotifications/FakebookNotifications.WebApi/Controller/NotificationController.cs]: "use the paramName argument to this constructor as well"
previous review comment not addressed

**Review by escalonn — CHANGES_REQUESTED** [review]

**Comment by ibnetahir1** [comment]
@escalonn I did address that change. The throw statement is moved to the switch expression and it does use the paramName argument.

File: FakebookNotifications/FakebookNotifications.WebApi/Controller/NotificationController.cs [hunk 19]
```diff
@@ -83,30 +87,19 @@ private async Task<IActionResult> CreateNotificationAsync(string loggedInUser, s
...
+            string methodName = notificationType switch
             {
-                throw new ArgumentException("notificationType is not valid");
-            }
-
+                NotificationType.Comment => "CommentNotification",
+                NotificationType.Like => "LikeNotification",
+                NotificationType.Follow => "FollowNotification",
+                _ => throw new ArgumentException("notificationType is not valid", nameof(notificationType))
+            };
```
Comment (ibnetahir1) [inline FakebookNotifications/FakebookNotifications.WebApi/Controller/NotificationController.cs:102]: this is where the throw statement is now.

**Review by escalonn — APPROVED** [review]
you're right, i had the review interface set wrong
